include/pmacc/particles/operations/CountParticles.hpp

```diff
@@ -67,75 +68,46 @@ namespace pmacc
             constexpr uint32_t numWorkers = T_numWorkers;
 
-            PMACC_SMEM(acc, frame, FramePtr);
             PMACC_SMEM(acc, counter, int);
-            PMACC_SMEM(acc, particlesInSuperCell, lcellId_t);
-
-            using SuperCellSize = typename T_Mapping::SuperCellSize;
-
-            DataSpace<dim> const threadIndex(cupla::threadIdx(acc));
-            auto const workerIdx
-                = static_cast<uint32_t>(DataSpaceOperations<dim>::template map<SuperCellSize>(threadIndex));
 
+            uint32_t const workerIdx = cupla::threadIdx(acc).x;
             DataSpace<dim> const superCellIdx(mapper.getSuperCellIndex(DataSpace<dim>(cupla::blockIdx(acc))));
 
             auto onlyMaster = lockstep::makeMaster(workerIdx);
 
-            onlyMaster(
-                [&]()
-                {
-                    frame = pb.getLastFrame(superCellIdx);
-                    particlesInSuperCell = pb.getSuperCell(superCellIdx).getSizeLastFrame();
-                    counter = 0;
-                });
+            auto forEachParticle
+                = pmacc::particles::algorithm::acc::makeForEach<numWorkers>(workerIdx, pb, superCellIdx);
+
+            onlyMaster([&]() { counter = 0; });
 
             cupla::__syncthreads(acc);
 
-            if(!frame.isValid())
-                return; // end kernel if we have no frames
+            // end kernel if we have no particles
+            if(!forEachParticle.hasParticles())
+                return;
             filter.setSuperCellPosition((superCellIdx - mapper.getGuardingSuperCells()) * mapper.getSuperCellSize());
 
             auto accParFilter = parFilter(
                 acc,
                 superCellIdx - mapper.getGuardingSuperCells(),
                 lockstep::Worker<numWorkers>{workerIdx});
 
-            auto forEachParticle = lockstep::makeForEach<frameSize, numWorkers>(workerIdx);
-
-            while(frame.isValid())
-            {
-                forEachParticle(
-                    [&](uint32_t const linearIdx)
-                    {
-                        if(linearIdx < particlesInSuperCell)
-                        {
-                            bool const useParticle = filter(*frame, linearIdx);
-                            if(useParticle)
-                            {
-                                auto parSrc = (frame[linearIdx]);
-                                if(accParFilter(acc, parSrc))
-                                    kernel::atomicAllInc(acc, &counter, ::alpaka::hierarchy::Threads{});
-                            }
-                        }
-                    });
-
-                cupla::__syncthreads(acc);
-
-                onlyMaster(
-                    [&]()
+            forEachParticle(
+                acc,
+                [&counter, &filter, &accParFilter](auto const& accelerator, auto& particle)
+                {
+                    bool const useParticle = filter(particle);
+                    if(useParticle)
```

Comment:
I am surprised, how the change in the construction of `forEachParticle` in line 81, changes its functor call signature.
We can discuss this offline tomorrow.